### ISSUE TITLE
Trivy code shows some vulnerabilties by Trivy Scanner itself

### ISSUE THREAD

**Issue by hideme4u** [aquasecurity/trivy#1439]
## Description
Trivy source code when scan using Trivy binary is showing few Vulnerabilities in it.

## What did you expect to happen?
In Trivy scanner code there should not be any vulnerabilities.

## What happened instead?
Trivy scanner showed vulnerabilities in it's own code base.

## Output of run with `-debug`:

```
(paste your output here)
```

## Output of `trivy -v`:
trivy -v
Version: 0.20.0
Vulnerability DB:

```
(paste your output here)
```
go.sum (gomod)
==============
Total: 22 (UNKNOWN: 2, LOW: 0, MEDIUM: 12, HIGH: 8, CRITICAL: 0)

+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
|              LIBRARY               | VULNERABILITY ID | SEVERITY |         INSTALLED VERSION         |             FIXED VERSION             |                  TITLE                  |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/apache/thrift           | CVE-2019-0205    | HIGH     | 0.12.0                            | 0.13.0                                | thrift: Endless loop when               |
|                                    |                  |          |                                   |                                       | feed with specific input data           |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2019-0205    |
+                                    +------------------+          +                                   +                                       +-----------------------------------------+
|                                    | CVE-2019-0210    |          |                                   |                                       | thrift: Out-of-bounds read              |
|                                    |                  |          |                                   |                                       | related to TJSONProtocol                |
|                                    |                  |          |                                   |                                       | or TSimpleJSONProtocol                  |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2019-0210    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-13949   |          |                                   | v0.14.0                               | libthrift: potential DoS when           |
|                                    |                  |          |                                   |                                       | processing untrusted payloads           |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-13949   |
+------------------------------------+------------------+          +-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/buger/jsonparser        | CVE-2020-10675   |          | 0.0.0-20180808090653-f4dd9f5a6b44 | v0.0.0-20200321185410-91ac96899e49    | golang-github-buger-jsonparser:         |
|                                    |                  |          |                                   |                                       | infinite loop via a Delete call         |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-10675   |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-35381   |          |                                   | v1.1.1                                | jsonparser: GET call can lead to        |
|                                    |                  |          |                                   |                                       | a slice bounds out of range...          |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-35381   |
+------------------------------------+------------------+          +-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/dgrijalva/jwt-go        | CVE-2020-26160   |          | 3.2.0+incompatible                |                                       | jwt-go: access restriction              |
|                                    |                  |          |                                   |                                       | bypass vulnerability                    |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-26160   |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/gorilla/handlers        | GO-2020-0020     | UNKNOWN  | 0.0.0-20150720190736-60c7bfde3e33 | v1.3.0                                |                                         |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/miekg/dns               | CVE-2019-19794   | MEDIUM   | 1.0.14                            | v1.1.25-0.20191211073109-8ebf2e419df7 | golang-github-miekg-dns: predictable    |
|                                    |                  |          |                                   |                                       | TXID can lead to response forgeries     |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2019-19794   |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/sassoftware/go-rpmutils | CVE-2020-7667    | HIGH     | 0.0.0-20190420191620-a8f1baeba37b | v0.1.0                                | In package                              |
|                                    |                  |          |                                   |                                       | github.com/sassoftware/go-rpmutils/cpio |
|                                    |                  |          |                                   |                                       | before version 0.1.0, the               |
|                                    |                  |          |                                   |                                       | CPIO extraction functionality           |
|                                    |                  |          |                                   |                                       | doesn't sanitize...                     |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-7667    |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| github.com/satori/go.uuid          | GO-2020-0018     | UNKNOWN  | 1.2.0                             | v1.2.1-0.20181016170032-d91630c85102  |                                         |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+
| k8s.io/kubernetes                  | CVE-2020-8558    | HIGH     | 1.13.0                            | 1.16.11, 1.17.7, 1.18.4               | kubernetes: node localhost services     |
|                                    |                  |          |                                   |                                       | reachable via martian packets           |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8558    |
+                                    +------------------+----------+                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2019-1002101 | MEDIUM   |                                   | 1.11.9, 1.12.7, 1.13.5,               | kubernetes: Mishandling of              |
|                                    |                  |          |                                   | 1.14.1-beta.0                         | symlinks allows for arbitrary           |
|                                    |                  |          |                                   |                                       | file write via `kubectl cp`...          |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2019-1002101 |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2019-11250   |          |                                   | v1.16.0-beta.1                        | kubernetes: Bearer tokens               |
|                                    |                  |          |                                   |                                       | written to logs at high                 |
|                                    |                  |          |                                   |                                       | verbosity levels (>= 7)...              |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2019-11250   |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8552    |          |                                   |                                       | kubernetes: Use of                      |
|                                    |                  |          |                                   |                                       | unbounded 'client' label in             |
|                                    |                  |          |                                   |                                       | apiserver_request_total allows          |
|                                    |                  |          |                                   |                                       | for memory exhaustion...                |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8552    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8554    |          |                                   |                                       | kubernetes: MITM using                  |
|                                    |                  |          |                                   |                                       | LoadBalancer or ExternalIPs             |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8554    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8555    |          |                                   | 1.15.11, 1.16.9, 1.17.5,              | kubernetes: Server side                 |
|                                    |                  |          |                                   | 1.18.1                                | request forgery (SSRF) in               |
|                                    |                  |          |                                   |                                       | kube-controller-manager                 |
|                                    |                  |          |                                   |                                       | allows users to leak...                 |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8555    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8557    |          |                                   | 1.16.13, 1.17.9, 1.18.6               | kubernetes: Node disk DOS by            |
|                                    |                  |          |                                   |                                       | writing to container /etc/hosts         |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8557    |
+                                    +------------------+          +                                   +                                       +-----------------------------------------+
|                                    | CVE-2020-8559    |          |                                   |                                       | kubernetes: compromised node could      |
|                                    |                  |          |                                   |                                       | escalate to cluster level privileges    |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8559    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8563    |          |                                   | 1.19.3                                | kubernetes: Secret leaks                |
|                                    |                  |          |                                   |                                       | in kube-controller-manager              |
|                                    |                  |          |                                   |                                       | when using vSphere Provider             |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8563    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8564    |          |                                   | v1.20.0-alpha.1                       | kubernetes: Docker config               |
|                                    |                  |          |                                   |                                       | secrets leaked when file is             |
|                                    |                  |          |                                   |                                       | malformed and loglevel >=...            |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8564    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2020-8565    |          |                                   | v1.20.0-alpha.2                       | kubernetes: Incomplete fix              |
|                                    |                  |          |                                   |                                       | for CVE-2019-11250 allows for           |
|                                    |                  |          |                                   |                                       | token leak in logs when...              |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2020-8565    |
+                                    +------------------+          +                                   +---------------------------------------+-----------------------------------------+
|                                    | CVE-2021-25735   |          |                                   | 1.18.18, 1.19.10, 1.20.6              | kubernetes: Validating                  |
|                                    |                  |          |                                   |                                       | Admission Webhook does not              |
|                                    |                  |          |                                   |                                       | observe some previous fields            |
|                                    |                  |          |                                   |                                       | -->avd.aquasec.com/nvd/cve-2021-25735   |
+------------------------------------+------------------+----------+-----------------------------------+---------------------------------------+-----------------------------------------+

**Comment by hideme4u** [comment]
@afdesk @masahiro331 @knqyf263  How Scanner code itself has vulnerabilities that should be fixed?

**Comment by afdesk** [comment]
@hideme4u thanks for your report!
I'm checking it

**Comment by afdesk** [comment]
I see problems with next packages: `github.com/apache/thrift`, `github.com/buger/jsonparser`, `github.com/dgrijalva/jwt-go`, `github.com/gorilla/handlers`, `github.com/miekg/dns`, `github.com/sassoftware/go-rpmutils`, `github.com/satori/go.uuid` and `k8s.io/kubernetes`.

but the main module does not need these modules. it seems they're used for tests.

**Comment by hideme4u** [comment]
@afdesk  Thanks for Checking.  But how do you verified they are used only for test not in main module?
Can you please help?

**Comment by afdesk** [comment]
@hideme4u oh, maybe i missed something. give me some time

**Comment by hideme4u** [comment]
@afdesk are you fixing this?  Please let us know.  Thanks for the reply.

**Comment by afdesk** [comment]
@hideme4u yes, I'm investigating it.
sure, I'll let you know about the results soon.

**Comment by afdesk** [comment]
@hideme4u 
thanks for your report about vulnerabilities in `trivy`.
I investigated these vulns, and our team discussed the result.

at first, `trivy` doesn't directly use modules with vulnerabilities.

but there are dependencies which contain other dependencies with vulnerabilities.
the some of these dependencies don't completely affect us.
the another part is known and takes a time to change it. 
we'll do it shortly.

**Comment by afdesk** [comment]
we investigated more this issue.

1. at first, I've run `trivy` without examples and integration tests:
```sh
$ trivy -d fs --skip-dirs integration/ --skip-dirs examples/ .
```
and found 5 packages with vulnerability: `github.com/apache/thrift`, `github.com/dgrijalva/jwt-go`, `github.com/miekg/dns`, `github.com/satori/go.uuid` and `k8s.io/kubernetes`.

. I've built the last version of `trivy`:
```sh
$ go build cmd/trivy/main.go
```
2. and then, I've checked new `./main`:
```sh
$ go version -m ./main | grep thrift
$ go version -m ./main | grep dgrijalva
$ go version -m ./main | grep dns
$ go version -m ./main | grep satori
$ go version -m ./main | grep kubernetes
```
`trivy` executable files doesn't contain dependencies with vulnerabilities.

maybe should we close this issue? ;)
